### PR TITLE
Add Support for the VC Data Model

### PR DESCRIPTION
**Pull request by OR13**
Attempting to address: https://github.com/decentralized-identity/.well-known/issues/25

This PR Adds tests for supporting for the JWT Proof format of the VC Data Model.

I used did:ethr and https://github.com/decentralized-identity/did-jwt-vc

To demonstrate the VC Data Model format.

I have also included support for Linked Data Proofs, because it was trivial to show how they can be supported.

No spec changes have been made, and no changes to the demo in its current form have been made.

The security context of the JWT Proof Variant of VC Data Model is the same as it was before. No JSON-LD processing is occurring (in fact there appears to be very little validation of the vc data structure performed by did-jwt-vc).

It appears that the spec language would need to be revised to align with the VC Data Model, further discussion is required.

### CODE REVIEW

File: contexts/did-configuration-v0.0.jsonld [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "@context": [
+    "https://schema.org",
+    {
+      "@version": 1.1,
+      "didcfg": "https://identity.foundation/.well-known/contexts/did-configuration-v0.0#",
+      "DomainLinkageAssertion": "didcfg:DomainLinkageAssertion",
```
[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: This was raised as an issue of the CCG mailing list regarding the vc greeting card example. Apparently the type of the credential should be defined in the context... I was also confused by this.

File: contexts/did-configuration-v0.0.jsonld [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "@context": [
+    "https://schema.org",
```
[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: It will when requested with the correct content headers I think... I pulled this from the did core spec: https://github.com/w3c/did-core/blob/master/contexts/did-v0.11.jsonld#L8

**Comment by OR13** [comment]
@awoie I don't think the examples you provided would be an issue, they are certainly more compact. The context can only provide definitions for things underneath it, so there would be no semantic definition for `sub` and `iss`... but I'm not sure we need that, the vc data model would probably have discussed this in depth, there most be a reason they chose the structure we have today.

File: contexts/did-configuration-v0.0.jsonld (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjgd) [reply contexts/did-configuration-v0.0.jsonld]: This is not the same thing.

In JSON-LD, if the @context property may be
- a map for specifying context definition
- an IRI to specify the location of another context
- an array combining either of the above.

- In this file, the https://schema.org url is the first element of the @context array and is not a map, therefore it should resolve to a valid context. See https://w3c.github.io/json-ld-syntax/#the-context

- In the file you link, the "schema" value is used for [IRI expansion](https://w3c.github.io/json-ld-syntax/#iri-expansion-within-a-context).

File: contexts/did-configuration-v0.0.jsonld (hunk 1 above)
[the remark this answers is not in the document]
Comment (gjgd) [reply contexts/did-configuration-v0.0.jsonld]: Could you link the supporting evidence for that? I'm not convinced because this seems in contradiction with what a context file is (mapping terms with IRIs)

[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: Here is the thread where this was discussed: https://lists.w3.org/Archives/Public/public-credentials/2019Dec/0006.html

You can also find similar examples here: https://www.w3.org/2018/credentials/examples/v1

For the example: https://www.w3.org/TR/vc-data-model/#example-1-a-simple-example-of-a-verifiable-credential

Although this seems to support your assertion since "AlumniCredential" is not present in the context.

File: contexts/did-configuration-v0.0.jsonld (hunk 2 above)
[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: hmm, I can see that schema.org is indeed being resolved by the document loader in the linked data proof example, here is a snippet that I pulled from the document loader:

```
https://schema.org { contextUrl: null,
      documentUrl: 'https://schema.org/docs/jsonldcontext.jsonld',
      document:
```

That url does resolve to a context... however, we are not using any schema.org properties in this proof. so I should be able to remove it, then use a property and see the test fail, then add it back again.

[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: I was able to confirm that the context is indeed being used by adding the following to the domain link assertion:

```
 "@type": "WebSite",
        url: "https://identity.foundation.com/"
```

An error occurs when schema.org url is omited, and the url property above is included.

[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: I think we should remove schema.org since we are not using it here.

File: contexts/did-configuration-v0.0.jsonld (hunk 1 above)
[the remark this answers is not in the document]
Comment (OR13) [reply contexts/did-configuration-v0.0.jsonld]: I think you are right, and I'm not seeing any error with the credential format, with "DomainLinkageAssertion" removed from the context.

File: contexts/did-configuration-v0.0.jsonld (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjgd) [reply contexts/did-configuration-v0.0.jsonld]: Interesting, I wonder how the "https://schema.org/" url gets translated to the full path that contains the json-ld "https://schema.org/docs/jsonldcontext.jsonld"

**Comment by csuwildcat** [comment]
Why can't we just leave the existing top-level properties and current validation rules, then specify that the `domain` property must also be present inside the VC portion to ensure validating it that way also works? How is that not valid?

**Comment by awoie** [comment]
> @awoie I don't think the examples you provided would be an issue, they are certainly more compact. The context can only provide definitions for things underneath it, so there would be no semantic definition for `sub` and `iss`... but I'm not sure we need that, the vc data model would probably have discussed this in depth, there most be a reason they chose the structure we have today.

@OR13 The examples for LD-Proofs and JWT are semantically equivalent. By following the W3C VC spec one will be able to construct a JSON-LD credential from the JWT. So, I would be in favour of using my examples in the spec.

**Comment by OR13** [comment]
Screen Shot 2020-01-31 at 3 31 43 PM

**Comment by OR13** [comment]
I have added tests for creating and verifying that a given did is linked via the configuration. 

I think the next step is to merge this demo and then update the spec to reflect the VC Data Model as is.

**Comment by awoie** [comment]
@OR13 why was the nesting of `domainLinkageAssertion` necessary? Isn't that the point of JSON-LD to avoid name clashes, so it would have been save to remove that nesting?

**Comment by OR13** [comment]
@awoie good question, I have opened this issue to gather feedback: https://github.com/decentralized-identity/.well-known/issues/29

TL;DR; i made it that way because I  thought there might at some point in the future be other links of assertions in the credential subject... but I'm happy to make the credential look however we want.
